File: Data Structures & Algorithms/trapping-rain-water/submission-0.java
class Solution {
    public int trap(int[] height) {
        int maxLeftValue = 0;
        int maxRightValue = 0;
        int[] maxLeft = new int[height.length];
        int[] maxRight = new int[height.length];
        // get every index's max on left and right
        for(int i=0;i<height.length;i++){
            if(i==0){
                maxLeft[i] = 0;
                continue;
            }
            maxLeftValue = Math.max(maxLeftValue, height[i-1]);
            maxLeft[i] = maxLeftValue;
        }
        for(int i=height.length-1;i>=0;i--){
            if(i==(height.length-1)){
                maxRight[i] = 0;
                continue;
            }
            maxRightValue = Math.max(maxRightValue, height[i+1]);
            maxRight[i] = maxRightValue;
        }
        // cal every index's area
        int res = 0;
        for(int i=0;i<height.length;i++){
            int newArea = Math.min(maxLeft[i], maxRight[i])-height[i];
            if( newArea>0 ){
                res+=newArea;
            }
        }
        return res;
    }
}

File: Data Structures & Algorithms/insert-new-interval/submission-0.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        res = []
        i = 0
        n = len(intervals)
        while i<n and intervals[i][1]<newInterval[0]:
            res.append(intervals[i])
            i+=1
        # overlap
        while  i<n and intervals[i][0]<=newInterval[1]:
            newInterval[0] = min(newInterval[0], intervals[i][0])
            newInterval[1] = max(newInterval[1], intervals[i][1])
            i+=1
        
        res.append(newInterval)
        
        # while i<n and intervals[i][0]>newInterval[1]:
        while i<n:
            res.append(intervals[i])
            i+=1

        # for i in range(len(intervals)):
        #     if inserted:
        #         if intervals[i][1]<newInterval[0]:
        #             res.append(intervals[i])
        #         elif intervals[i][0]>newInterval[1]:
        #             res.append(intervals[i])
        #     else:
        #         if intervals[i][1]<newInterval[0]:
        #             res.append(intervals[i])
        #         else:
        #             newInterval[0] = min(newInterval[0], intervals[i][0])
        #             newInterval[1] = max(newInterval[1], intervals[i][1])
        #             if i+1<len(intervals) and intervals[i+1][0]>newInterval[1]:
        #                 res.append(newInterval)
        #                 inserted = True
                
        return res

